roachprod: enable roachprod to be configured for external use

ajwerner opened the pull request:
After this change, roachprod fully works on GCP at my company with the following configuration:

```
ROACHPROD_DNS_REQUIRED_PROVIDERS=gce
ROACHPROD_EMAIL_DOMAIN=@dataexmachina.dev
ROACHPROD_GCE_DEFAULT_PROJECT=dem-ephemeral
ROACHPROD_GCE_DEFAULT_SERVICE_ACCOUNT=roachprod@dem-ephemeral.iam.gserviceaccount.com
ROACHPROD_GCE_DNS_DOMAIN=roachprod.exray.cloud
ROACHPROD_GCE_DNS_MANAGED_DOMAIN=roachprod-managed.dataexmachina.dev
ROACHPROD_GCE_DNS_MANAGED_ZONE=roachprod-managed
ROACHPROD_GCE_DNS_PROJECT=dem-ephemeral
ROACHPROD_GCE_DNS_ZONE=roachprod
```

All of these configuration options can also be set via flags.

See individual commits.

Epic: none
Release note: none

ajwerner committed:
roachprod: make email domain configurable

This commit introduces the email domain flag persistently and sets its
default from the environment using ROACHPROD_EMAIL_DOMAIN.

Epic: None

Release note: None

ajwerner committed:
roachprod: allow configuration of gce

This allows both flag and environment variable configuration for a
variety of critical pieces of functionality.

Epic: None

Release note: None

ajwerner committed:
roachprod: apply provider configure flags more

These subcommands also benefit from this configuration.

Epic: None

Release note: None

ajwerner committed:
roachprod: configure providers needed for DNS

Before this change, DNS would only sync if the AWS provider was active.
This is now configurable.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120340)
<!-- Reviewable:end -->

srosenberg commented on pkg/cmd/roachprod/flags.go:
```diff
@@ -131,7 +133,7 @@ func initFlags() {
 
 			for _, cmd := range []*cobra.Command{
-				destroyCmd, extendCmd, listCmd, syncCmd, gcCmd,
+				destroyCmd, extendCmd, listCmd, syncCmd, gcCmd, setupSSHCmd, startCmd, pgurlCmd, adminurlCmd,
```
Methinks only `setupSSHCmd` calls `Sync`, which in turn may invoke `SyncDNS`. The others don't need parameterized DNS flags.

srosenberg reviewed (approved):
Thanks for refactoring hardcoded configs and making `roachprod` useful externally! I played around with the PR locally; things appear to work as they should. Still, it wouldn't hurt to do a full run of the nightlies as a sanity check.

@renatolabs @herkolategan I'd appreciate another pair of eyes.

andreimatei commented:
I rebased the patches. A review would be greatly appreciated; this PR will help us do other good things for cockroach :)
In particular, I want to propose some `roachtest` patches and getting `roachprod` working locally in an easier way would be great.

srosenberg commented:
> I rebased the patches. A review would be greatly appreciated; this PR will help us do other good things for cockroach :) In particular, I want to propose some `roachtest` patches and getting `roachprod` working locally in an easier way would be great.

Thanks and sorry for the delay! I am happy with the changes. The reason it probably stalled is we didn't do any smoke testing. Lemme do some local testing as well as kick off a nightly run with `SELECT_PROBABILITY`.

srosenberg commented:
Will check the smoke test, [SELECT_PROBABILITY=0.6](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15369312?buildTab=log&logFilter=debug&focusLine=5&logView=linear), when it completes.

andreimatei commented:
Ugh, I think there might be something broken, so please hold on on the testing. I think things are good when `roachprod` is used as a binary, but not when `roachtest` uses it as a library. I'm working on `roachtest` myself, so I noticed.
I'll ping back when I've figured it out. Sorry for the noise.

srosenberg commented:
> Ugh, I think there might be something broken, so please hold on on the testing. I think things are good when `roachprod` is used as a binary, but not when `roachtest` uses it as a library. I'm working on `roachtest` myself, so I noticed. I'll ping back when I've figured it out. Sorry for the noise.

No worries! Yep, seeing a lot of these in CI, 

```
ERROR: (gcloud.compute.project-info.describe) The project property is set to the empty string, which is invalid.
```

andreimatei committed:
roachprod: permit disabling of Prometheus registration

This patch makes an empty value for the COCKROACH_PROM_HOST_URL env var
disable the roachprod's cluster registration with the Prometheus
registration service. This is useful for people who don't work at
CockroachLabs and don't have access to this service.

Release note: None
Epic: None

andreimatei commented:
OK, I believe I've fixed the issue. roachtests run fine for me now.
I've also had to pile on one more commit to optionally disable the integration with the Prometheus registration service. PTAL
Thanks!

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 6 files at r6, 2 of 3 files at r8, 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120340#-:-Nyb7l7M2hg1QA7IgxgS:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @DarrylWong, @herkolategan, and @renatolabs)



<!-- Sent from Reviewable.io -->

srosenberg commented:
Smoke test: [SELECT_PROBABILITY=0.6](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15383971?buildTab=log&logFilter=debug&focusLine=5&logView=linear).

andreimatei commented:
I've pushed the fix to the problem I was describing in https://github.com/cockroachdb/cockroach/pull/124678#issuecomment-2136040318.
If you kick off another smoke test, I'll verify in the runner logs that the cluster reuse is working. (a more limited run of only a few tests, for example `roachtest run acceptance`, should do if that's easy).
Otherwise, the previous run looked fine, right?
Thanks!

srosenberg commented:
Kicked off another run of [SELECT_PROBABILITY=0.6](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15443185?buildTab=log&logFilter=debug&focusLine=6&logView=linear).

andreimatei commented:
Cluster reuse looks good now -- you see `Using existing cluster` in the runner's logs.

srosenberg commented:
Smoke test results look good! Resolving merge conflict and merging. Thanks for making roachprod ready for external use!

srosenberg commented:
Final smoke test ([SELECT_PROBABILITY=0.25](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15462087?buildTab=log&logFilter=debug&focusLine=6&logView=linear&linesState=3) since there were non-trivial merge conflicts after   [1] got merged.

[1] https://github.com/cockroachdb/cockroach/pull/124604

srosenberg commented:
Final smoke test looks good, let's 🚢 

bors r+
